test/writeasync_test.dart

```diff
@@ -0,0 +1,244 @@
+    await transaction.commitAsync();
+
+    expect(realm.all<Person>().length, 1);
+  });
+
+  test('Realm.beginWriteAsync rollback undoes changes', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final transaction = await realm.beginWriteAsync();
+    realm.add(Person('John'));
+    transaction.rollback();
+
+    expect(realm.all<Person>().length, 0);
+  });
+
+  test('Realm.writeAsync allows persists changes', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    await realm.writeAsync(() {
+      realm.add(Person('John'));
+    });
+
+    expect(realm.all<Person>().length, 1);
+  });
+
+  test('Realm.beginWriteAsync when realm is closed undoes changes', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+
+    await realm1.beginWriteAsync();
+    realm1.add(Person('John'));
+    realm1.close();
+
+    expect(realm2.all<Person>().length, 0);
+  });
+
+  test("Realm.writeAsync with multiple transactions doesnt't deadlock", () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final t1 = await realm.beginWriteAsync();
+    realm.add(Person('Marco'));
+
+    final writeFuture = realm.writeAsync(() {
+      realm.add(Person('Giovanni'));
+    });
+
+    await t1.commitAsync();
+    await writeFuture;
+
+    final people = realm.all<Person>();
+    expect(people.length, 2);
+    expect(people[0].name, 'Marco');
+    expect(people[1].name, 'Giovanni');
+  });
+
+  test('Realm.writeAsync returns valid objects', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final person = await realm.writeAsync(() {
+      return realm.add(Person('John'));
+    });
+
+    expect(person.name, 'John');
+  });
+
+  test('Realm.writeAsync throws user exception', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    try {
+      await realm.writeAsync(() {
+        throw Exception('User exception');
+      });
+    } on Exception catch (e) {
+      expect(e.toString(), 'Exception: User exception');
+    }
+  });
+
+  test('Realm.writeAsync FIFO order ensured', () async {
+    final acquisitionOrder = <int>[];
+    final futures = <Future<void>>[];
+
+    final realm = getRealm(Configuration.local([Person.schema]));
+
+    for (var i = 0; i < 5; i++) {
+      futures.add(realm.writeAsync(() {
+        acquisitionOrder.add(i);
+      }));
+    }
+
+    await Future.wait(futures);
+
+    expect(acquisitionOrder, [0, 1, 2, 3, 4]);
+  });
+
+  test('Realm.beginWriteAsync with cancellation token', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+    final t1 = realm1.beginWrite();
+
+    final token = CancellationToken();
+    Future<void>.delayed(Duration(milliseconds: 1)).then((value) => token.cancel());
```

Comment:
```suggestion
    final token = TimeoutCancellationToken(const Duration(milliseconds: 1), timeoutException: CancelledException());
```
Slightly shorter.. In case you didn't know about this?